MSMAll/scripts/SingleSubjectConcat.sh

```diff
@@ -1,534 +1,180 @@
+
+AtlasFolder="${StudyFolder}/${Subject}/MNINonLinear"
+log_Msg "AtlasFolder: ${AtlasFolder}"
+
+OutputFolder="${AtlasFolder}/Results/${OutputfMRIName}"
+log_Msg "OutputFolder: ${OutputFolder}"
+# create folder to save the concatenated file
+mkdir -p "${OutputFolder}"
+
+Caret7_Command=${CARET7DIR}/wb_command
+log_Msg "Caret7_Command: ${Caret7_Command}"
+
+# read frame informations
+runFrames=()
+runTR=()
+for ((index = 0; index < ${#fMRINamesArray[@]}; ++index))
+do
+	fMRIName="${fMRINamesArray[index]}"
+	ResultsFolder="${AtlasFolder}/Results/${fMRIName}"
+	log_Msg "ResultsFolder: ${ResultsFolder}"
+	FullDenseTCS=${ResultsFolder}/${fMRIName}${fMRIProcSTRING}.dtseries.nii
+	log_Msg "FullDenseTCS: ${FullDenseTCS}"
+	NumFrames=`${Caret7_Command} -file-information "${FullDenseTCS}" -only-number-of-maps`
+	runFrames[index]="$NumFrames"
+	TR=`${Caret7_Command} -file-information "${FullDenseTCS}" -only-step-interval`
+	runTR[index]="$TR"
+	if [ $((index)) -eq 0 ]; then
+		minFrames=${runFrames[index]}
 	else
-		log_Msg "Demean: ${p_Demean}"
+		(( ${runFrames[index]}<minFrames )) && minFrames=${runFrames[index]}
 	fi
+done
 
-	if [ -z "${p_VarianceNormalization}" ]; then
-		log_Err "Variance Normalization (--variance-normalization=<YES | NO>) required"
-		error_count=$(( error_count + 1 ))
-	else
-		log_Msg "Variance Normalization: ${p_VarianceNormalization}"
+EndFrameToUse=""
+if [[ "$EndFrame" == "" ]]; then
+	# preserving the timeseries to the end
+	FrameString="Frame${StartFrame}ToTheEnd"
+else
+	# check valid frame range
+	if ((StartFrame > minFrames ||  EndFrame > minFrames || StartFrame < 1 || EndFrame < 1 || StartFrame > EndFrame)); then
+		log_Err_Abort "The provided start frame ${StartFrame} and end frame ${EndFrame} is not valid, it must be between the range 1 to ${minFrames}. Please check the fMRI runs."
 	fi
+	FrameString="Frame${StartFrame}To${EndFrame}"
+fi
 
-	if [ -z "${p_ComputeVarianceNormalization}" ]; then
-		log_Err "Compute Variance Normalization (--compute-variance-normalization=<YES | NO> required"
-		error_count=$(( error_count + 1 ))
-	else
-		log_Msg "Compute Variance Normalization: ${p_ComputeVarianceNormalization}"
-	fi
+log_Msg "FrameString: ${FrameString}"
 
-	if [ -z "${p_RevertBiasField}" ]; then
-		log_Err "Revert Bias Field (--revert-bias-field=<YES | NO>) required"
-		error_count=$(( error_count + 1 ))
+duration=0
+for ((index = 0; index < ${#fMRINamesArray[@]}; ++index))
+do
+	TR="${runTR[index]}"
+	if [[ "$EndFrame" == "" ]]; then
+		Frames=$((runFrames[index]-StartFrame+1))
 	else
-		log_Msg "Revert Bias Field: ${p_RevertBiasField}"
+		Frames=$((EndFrame-StartFrame+1))
 	fi
+	duration=$(bc -l <<< $duration+$TR*$Frames)
+done
+
+# convert to mins secs
+m=$(bc <<< "${duration}/60")
+s=$(bc <<< "${duration}%60")
+DurationString="${m}mins${s}secs"
+
+MergeArray=()
+cnt=0
+# loop for demean+vn+concat
+for ((index = 0; index < ${#fMRINamesArray[@]}; ++index)) ; do
+    fMRIName="${fMRINamesArray[$index]}"
```

Comment:
This line uses space for indentation, the rest use tabs (except for the stuff from the `newopts.shlib` template).  Please try not to mix indentation types in consecutive lines, at least.  Personally, I prefer 4 spaces for indentation, but that isn't standard across the pipelines.

Reply:
I usually use two spaces in shell and a tab in Matlab.